### PR TITLE
Create form field number

### PR DESCRIPTION
**Pull request by Devessier**
- Refactor VariableTagInput to have a reusable low-level TipTap editor
- Create three primitive form fields:
  - Text
  - Number
  - Boolean

## Notes

- We should automatically recognize the placeholder to use for every FormFieldInput, as it's done for FieldInputs.

## Design decisions

Our main challenge was for variables and inputs to be able to communicate between each other. We chose an API that adds some duplication but remains simple and doesn't rely on "hacks" to work. Common styles are centralized.

## Demo

"Workflow" mode with variables:

CleanShot 2024-11-26 at 10 43 25@2x

FormFieldInput mode, without variables:

CleanShot 2024-11-26 at 10 44 26@2x

Behavior difference between fields that can contain variables and static content, and inputs that can have either a variable value or a static value:

CleanShot 2024-11-26 at 10 47 13@2x

### CODE REVIEW

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx
Comment (thomtrp) [inline packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx:34]: multiline not needed for number

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx
Comment (thomtrp) [inline packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx:20]: Should be a common Overlay for primitive components

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx
Comment (thomtrp) [inline packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx:159]: don't we have a component already for this?

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx
Comment (thomtrp) [inline packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx:147]: let's separate:
- FormNumberFieldInput should be a simple number input with a FormOverlay to add design
- Variable button / dropdown should come from the parent

File: packages/twenty-front/src/modules/workflow/components/WorkflowFormTextFieldInput.tsx
Comment (Devessier) [inline packages/twenty-front/src/modules/workflow/components/WorkflowFormTextFieldInput.tsx:33]: In order to minimize the number of times we do these time consuming calls, we might want to do these lazily:

```suggestion
      onPersist(() => {
        const jsonContent = editor.getJSON();
        const parsedContent = parseEditorContent(jsonContent);
      });
```

**Comment by martmull** [comment]
Functional testing observations
- Boolean, we should be able to set no value at all, currently we are forced to set true or false
- Number, we should not be able to type "string" values in input, only numbers
image
- Number, I think we should not allow float values either
image

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx:42]: Should we persist every time we change or could it be triggered by onBlur for example ? Or enter / esc ? This is maybe more product related.

**Comment by thomtrp** [comment]
> * Boolean, we should be able to set no value at all, currently we are forced to set true or false (we might update the `<BooleanFieldInput />` component) @Bonapara what do you think?

@martmull this is a discussion we got 👍  This is the current behavior of our Boolean field, this is why we kept the same. It should not be seen as a primitive boolean field but more as a True/False field. If the user needs a 3 options field (like null, false, true) he can still use the select field for this.

File: packages/twenty-front/src/modules/object-record/record-field/form-types/hooks/useTextVariableEditor.ts
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/object-record/record-field/form-types/hooks/useTextVariableEditor.ts:54]: Could this conflict with our useScopedHotkeys ? It should be more appropriate to add a hotkey scope for this and call useScopedHotkeys here no ?

File: packages/twenty-front/src/modules/workflow/components/WorkflowFormFieldInputBase.tsx
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/workflow/components/WorkflowFormFieldInputBase.tsx:87]: Strange pattern, we don't have that, maybe a separate callback would be more self-explanatory ? No need to be too dry.

File: packages/twenty-front/src/modules/workflow/components/WorkflowFormTextFieldInput.tsx
Comment (lucasbordeau) [reply packages/twenty-front/src/modules/workflow/components/WorkflowFormTextFieldInput.tsx:33]: See my previous comment, is it necessary to call onPersist at each change ?

File: packages/twenty-front/src/modules/workflow/components/WorkflowFormTextFieldInput.tsx
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/workflow/components/WorkflowFormTextFieldInput.tsx:54]: Could we avoid writing handlers inline ? We make the effort to write them in a section in the component above the JSX part pretty much everywhere with a coherent naming : handleXYZ

File: packages/twenty-front/src/modules/workflow/search-variables/components/SearchVariablesDropdown.tsx
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/workflow/search-variables/components/SearchVariablesDropdown.tsx:35]: Shouldn't it be `onInsertVariableTag`  ?

File: packages/twenty-front/src/modules/workflow/search-variables/components/VariableTagInput.tsx
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/workflow/search-variables/components/VariableTagInput.tsx:63]: Write a handler ?

File: packages/twenty-front/src/modules/workflow/search-variables/utils/extractVariableLabel.ts
Comment (martmull) [inline packages/twenty-front/src/modules/workflow/search-variables/utils/extractVariableLabel.ts:5]: can we use `REGEX_VARIABLE_TAG` here? I don't get the functional difference between the 2 regex

**Review by martmull — COMMENTED** [review]
Hey nice job, some comments

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx
Comment (Devessier) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormNumberFieldInput.tsx:42]: Thanks for mentioning this, @lucasbordeau. Currently, we persist in every change with some debouncing in the middle. We asked ourselves whether we had to implement the same blur and key listeners as for FieldInputs. However, we thought it was unnecessary as workflows don't work that way yet. We can revisit that decision later.

**Review by thomtrp — COMMENTED** [review]
@Devessier I am not sure to get the full picture here. I was expecting an architecture like:
- Primitive input (Boolean, Number, Text) that would be a common base for Form Field Input and Field Input
- Primitive form field input (BooleanFormFieldInput, NumberFormFieldInput...)
- Global component FormFieldInput to switch on it
- WorkflowFormFieldInput that send the update functions, the variable component as right element and call the FormFieldInput as main input component.
- Hard to implement for tiptap, so we could have a special WorkflowFormFieldInputWithTipTap (not the real component name)

In your code:
- I do not see FormFieldInput actually used? To me it was supposed to be the base for workflow forms but also the futur forms without variables we will have to implement
- It seems that we will have to add a WorkflowFieldInput for each type. But except for text using tiptap, isn't the logic to store value and use variables always the same?

Happy to discuss the architecture, maybe this is unrealistic. I would also like your opinion @lucasbordeau @charlesBochet

File: packages/twenty-front/src/modules/object-record/record-field/form-types/hooks/useTextVariableEditor.ts
Comment (Devessier) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/hooks/useTextVariableEditor.ts:54]: I'm not sure how we can integrate hotkey scopes with tiptap here. I agree that it might be better to use hotkey scopes. We should see who we can do that. Maybe @thomtrp has already tried to implement that?

File: packages/twenty-front/src/modules/workflow/search-variables/utils/extractVariableLabel.ts
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/search-variables/utils/extractVariableLabel.ts:5]: This regex captures what's inside brackets. The other one also captures brackets. We could refactor it to use a single regex.

File: packages/twenty-front/src/modules/object-record/record-field/form-types/hooks/useTextVariableEditor.ts
Comment (lucasbordeau) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/hooks/useTextVariableEditor.ts:54]: If tip tap doesn't allow us to take precedence over its listeners, maybe we could at least set a hotkey scope and check with a simple if whether the hotkey scope is activated for this field ?

Comment (lucasbordeau) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/hooks/useTextVariableEditor.ts:54]: That way we avoid conflicts, especially on forms where there are multiple inputs listening at the same time

**Comment by Devessier** [comment]
> @Devessier I am not sure to get the full picture here. I was expecting an architecture like:
> 
> * Primitive input (Boolean, Number, Text) that would be a common base for Form Field Input and Field Input
> * Primitive form field input (BooleanFormFieldInput, NumberFormFieldInput...)
> * Global component FormFieldInput to switch on it
> * WorkflowFormFieldInput that send the update functions, the variable component as right element and call the FormFieldInput as main input component.
> * Hard to implement for tiptap, so we could have a special WorkflowFormFieldInputWithTipTap (not the real component name)
> 
> In your code:
> 
> * I do not see FormFieldInput actually used? To me it was supposed to be the base for workflow forms but also the futur forms without variables we will have to implement
> * It seems that we will have to add a WorkflowFieldInput for each type. But except for text using tiptap, isn't the logic to store value and use variables always the same?
> 
> Happy to discuss the architecture, maybe this is unrealistic. I would also like your opinion @lucasbordeau

I'm unsure how we could have a standalone `FormFieldInput` component that could also be enhanced in `WorkflowFormFieldInput`. I might miss something, but I am happy to consider doing things differently.

Let's take `WorkflowFormNumberFieldInput` as an example. When we want to unlink a variable, we must call `setDraftValue`. However, if the `WorkflowFormNumberFieldInput` component was relying on `FormNumberFieldInput` to work, and provided that `FormNumberFieldInput` would need to own `draftValue` and `setDraftValue` to work as a standalone component, the parent component would need a way to set the state of its child component.
We could solve this by putting the input's state on Recoil, but it only solves some problems.

The `FormNumberFieldInput` component renders the `FormFieldInputBase` component, which gives it a standalone component look. The `WorkflowFormNumberFieldInput` component does so with `WorkflowFormFieldInputBase`.

We could find solutions to reuse standalone components and enhance them with variables, but I wonder if this is the simplest way.
Instead, I suggest reusing base components and common logic. For instance, both the `WorkflowFormNumberFieldInput` and `FormNumberFieldInput` components rely on the `TextInput` component.
We could also extract the common logic in a hook. For instance, `draftValue`, `persistNumber` and `handleChange` are common.
I didn't implement that in my PR, but I think it would be the way to go once we understood the common logic for each component.

I prefer some duplication of templating code (like creating more components) to trying to make things work together and finding in two weeks that one component shouldn't work the same way in workflows as it does in usual forms. Furthermore, we don't yet have a use case for `FormFieldInput`.

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormBooleanFieldInput.tsx
Comment (Devessier) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormBooleanFieldInput.tsx:37]: It's an important subject, and we should discuss it first. There are two policies: either we cast values to the type we expect or discard values that don't match their expected type. By casting, I mean doing things like this: `String(mightBeAStringOrSomethingElse)`.

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormTextFieldInput.tsx
Comment (Devessier) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormTextFieldInput.tsx:21]: I'm not sure to understand why multiline wouldn't be used. It appears to be used multiple times in the commend for me. Am I missing something?

File: packages/twenty-front/src/modules/workflow/components/WorkflowFormNumberFieldInput.tsx
Comment (martmull) [inline packages/twenty-front/src/modules/workflow/components/WorkflowFormNumberFieldInput.tsx:56]: same value={`${...}`}

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormTextFieldInput.tsx
Comment (martmull) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormTextFieldInput.tsx:21]: By saying not used, i mean i don't see any component that set this multiline parameter to true:
``` 
<ComponentUsingMultilineTrue
...
multiline
...
/>
```
Did I miss one of them?

File: packages/twenty-front/src/modules/object-record/record-field/form-types/hooks/useTextVariableEditor.ts
Comment (Devessier) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/hooks/useTextVariableEditor.ts:54]: I created a separate issue to track that refactor: https://github.com/twentyhq/twenty/issues/8786.

File: packages/twenty-front/src/modules/workflow/components/WorkflowFormBooleanFieldInput.tsx
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowFormBooleanFieldInput.tsx]: We can improve this in another PR. However, it will require many changes, which we must consider beforehand.

File: packages/twenty-front/src/modules/workflow/components/WorkflowFormNumberFieldInput.tsx
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowFormNumberFieldInput.tsx]: As mentioned here, I think we can check it later: https://github.com/twentyhq/twenty/pull/8634#discussion_r1861861097.

File: packages/twenty-front/src/modules/workflow/components/WorkflowFormBooleanFieldInput.tsx
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowFormBooleanFieldInput.tsx]: Created another issue to track that: https://github.com/twentyhq/twenty/issues/8787.

File: packages/twenty-front/src/modules/workflow/components/WorkflowFormNumberFieldInput.tsx
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowFormNumberFieldInput.tsx]: I duplicated the logic of the NumberFieldInput file; I'm unsure if we want to behave differently than this component.

File: packages/twenty-front/src/modules/workflow/search-variables/utils/initializeEditorContent.ts
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/search-variables/utils/initializeEditorContent.ts]: Done

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/SelectedVariableChip.tsx
Comment (thomtrp) [inline packages/twenty-front/src/modules/object-record/record-field/form-types/components/SelectedVariableChip.tsx:15]: VariableChip as a name is enough I think. We have a RecordChip, that will be consistent!

File: packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx
Comment (thomtrp) [inline packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx:172]: 😍

**Review by thomtrp — APPROVED** [review]
Amazing work @Devessier!! 💪 🔥 

Let's wait for @lucasbordeau review but LGTM!

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/SelectedVariableChip.tsx
Comment (Devessier) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/components/SelectedVariableChip.tsx:15]: Great 👌

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormTextFieldInput.tsx
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormTextFieldInput.tsx:46]: Could we avoid using just "variable" alone, either `variableValue`, `variableContent` or `variableName` ? As a first-time reader I don't clearly understand, which for this "meta" part of the code is important.

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/SelectedVariableChip.tsx
Comment (Devessier) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/components/SelectedVariableChip.tsx]: Done

File: packages/twenty-front/src/modules/object-record/record-field/form-types/types/VariablePickerComponent.ts
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/object-record/record-field/form-types/types/VariablePickerComponent.ts:5]: variable

File: packages/twenty-front/src/modules/workflow/search-variables/components/SearchVariablesDropdown.tsx
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/workflow/search-variables/components/SearchVariablesDropdown.tsx:35]: variable

File: packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormTextFieldInput.tsx
Comment (Devessier) [reply packages/twenty-front/src/modules/object-record/record-field/form-types/components/FormTextFieldInput.tsx]: Yes 👌 Done.

**Pull request merged**
